iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,302 @@
+public class PropertyCollection : TwinCollection
+{
+}
+```
+</details>
+
+
+<details>
+<summary>WritableProperty</summary>
+
+```csharp
+/// <summary>
+/// Empty constructor.
+/// </summary>
+public WritablePropertyResponse() { }
+
+/// <summary>
+/// Convenience constructor for specifying the properties.
+/// </summary>
+/// <param name="propertyValue">The unserialized property value.</param>
+/// <param name="ackCode">The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.</param>
+/// <param name="ackVersion">The acknowledgement version, as supplied in the property update request.</param>
+/// <param name="ackDescription">The acknowledgement description, an optional, human-readable message about the result of the property update.</param>
+public WritablePropertyResponse(object propertyValue, int ackCode, long ackVersion, string ackDescription = null)
+{
+    PropertyValue = propertyValue;
+    AckCode = ackCode;
+    AckVersion = ackVersion;
+    AckDescription = ackDescription;
+}
+
+/// <summary>
+/// The unserialized property value.
+/// </summary>
+[JsonProperty("value")]
+public object PropertyValue { get; set; }
+
+/// <summary>
+/// The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.
+/// </summary>
+[JsonProperty("ac")]
+public int AckCode { get; set; }
+
+/// <summary>
+/// The acknowledgement version, as supplied in the property update request.
+/// </summary>
+[JsonProperty("av")]
+public long AckVersion { get; set; }
+
+/// <summary>
+/// The acknowledgement description, an optional, human-readable message about the result of the property update.
+/// </summary>
+[JsonProperty("ad", DefaultValueHandling = DefaultValueHandling.Ignore)]
+public string AckDescription { get; set; }
+```
+</details>
+
+
+### Telemetry
+
+```csharp
+/// <summary>
+/// Sends a single instance of telemetry.
+/// </summary>
+/// <param name="telemetryName">The name of the telemetry to send.</param>
+/// <param name="telemetryValue">The value of the telemetry to send.</param>
+/// <param name="conventionHandler">A convention handler that defines the content encoding and serializer to use for the telemetry message.</param>
+/// <param name="componentName">The component name this telemetry belongs to.</param>
+/// <param name="cancellationToken">A cancellation token.</param>
+/// <remarks>
+/// This will create a single telemetry message and will not combine multiple calls into one message. Use <seealso cref="SendTelemetryAsync(IDictionary{string, dynamic}, string, IConventionHandler, CancellationToken)"/>. Refer to the documentation for <see cref="IConventionHandler"/> if you want to use a custom serializer.
+/// </remarks>
+public Task SendTelemetryAsync(string telemetryName, object telemetryValue, string componentName = default, IConventionHandler conventionHandler = default, CancellationToken cancellationToken = default);
```

Comment:
- I would say that it is safe to ignore the "key-value" and IDict overloads of the telemetry API; I think those are just leftover ones that James probably missed removing.
- We will expose a single API which takes in a `Message`. Now, this `Message` can be constructed with some serialization and encoding customizations (for which we will provide a default implementation). I haven't put the component name into this "serialization and encoding customizations" class, because in my mind component is more of a "modelling" related concept than a telemetry payload related concept. So, in this API, the user would need to set the component name explicitly on the message.
Thoughts?

```csharp
public class TelemetryConvention
{
	private const string ApplicationJsonContentType = "application/json";

	/// <summary>
	///
	/// </summary>
	public static readonly TelemetryConvention Instance = new TelemetryConvention();

	/// <summary>
	///
	/// </summary>
	public Encoding ContentEncoding { get; set; } = Encoding.UTF8;

	/// <summary>
	///
	/// </summary>
	public string ContentType { get; set; } = ApplicationJsonContentType;

	/// <summary>
	///
	/// </summary>
	/// <param name="objectToSerialize"></param>
	/// <returns></returns>
	public virtual string SerializeToString(object objectToSerialize)
	{
		return JsonConvert.SerializeObject(objectToSerialize);
	}

	/// <summary>
	///
	/// </summary>
	/// <param name="contentPayload"></param>
	/// <returns></returns>
	public virtual byte[] EncodeStringToByteArray(string contentPayload)
	{
		return ContentEncoding.GetBytes(contentPayload);
	}

	/// <summary>
	///
	/// </summary>
	/// <param name="objectToSendWithConvention"></param>
	/// <returns></returns>
	public virtual byte[] GetObjectBytes(object objectToSendWithConvention)
	{
		return EncodeStringToByteArray(SerializeToString(objectToSendWithConvention));
	}

	/// <summary>
	///
	/// </summary>
	/// <param name="telemetryName"></param>
	/// <param name="telemetryValue"></param>
	/// <returns></returns>
	public static IDictionary<string, object> FormatTelemetryPayload(string telemetryName, object telemetryValue)
	{
		return new Dictionary<string, object> { { telemetryName, telemetryValue } };
	}
}
```

caller:
```csharp
string telemetryName = "deviceHealth";
var deviceHealth = new DeviceHealth
{
	Status = "Running",
	RunningTimeInSeconds = s_stopwatch.Elapsed.TotalSeconds,
	StopRequested = false,
	StartTime = s_applicationStartTime,
};

IDictionary<string, object> telemetryPayload = TelemetryConvention.FormatTelemetryPayload(telemetryName, deviceHealth);

var messageToSend = new Message(telemetryPayload, TelemetryConvention.Instance)
{
	Properties = { ["property1"] = "myValue" },
	ComponentName = componentName,
};

await _deviceClient.SendTelemetryAsync(messageToSend, cancellationToken);
```

Reply:
the virtual methods in TelemetryConvention can be overridden by the customer to provide their own serilization/ encoding logic.

Reply:
OK - I've been staring at this some and I think why we're not totally on same page RE where `component` lives is what we're doing with this "TelemetryConventionThingy" isn't 100% aligned.

In C# if I'm reading correctly, you want this to tie in as a data serializer - in addition to being able to set messageProperties on the TelemetryMessage object as a convenience.  Correct?

C on the other hand is much more limited.  We have the telemetryAttributes thing to help customers "fall into the pit of success" by showing a bunch of parameters that are highly-desirable to be specified.  We are NOT in the serialization business and don't plan to be; the app needs to serialize their body before coming to us.

I'm wondering though if we should be out of this business completely?
So on C, we'd remove telemetryAttributes stuff from proposed API.  Just document the already-existing Message "accessors" (not technically correct but it maps into C# concept) and also add a new "accessor" for ComponentName.  So it becomes

```c
IOTHUB_MESSAGE_HANDLE h = IoTHub_Message_CreateFromString("{some-json-stuff}")
IoTHubMessage_SetEncoding(h,"...); // pre-existing
IoTHubMessage_SetContentType(h, ...);  // pre-existing
IoTHubMessage_SetComponentName(h,...); // would be new.  Right now in summer refresh samples apps need to know that this is "$.sub" and that's only MQTT friendly, so that's lame, but an accesor would "do right"
```

The message is fairly well understood construct already and not hard for a new person to IoT Hub to get head around as this is basic in any messaging framework.  So on C at least I don't see value of the telemtyrAttributes, or if we think it's a good idea we could just add later.

Could C# follow similar pattern?  FWIW (de)serialization of JSON properties is huge value SDK adds, but I think just translating a C# object using JSON serialization the bang-for-bug isn't as big because C# has native patterns around this as-is.  And then if the focus becomes on setting message properties - well add Component as 1st class citizen and document other stuff.

?

Reply:
Yup, that's exactly what I meant.

FWIW componentName isn't a top-level property on Message in C, so if C# is doing right thing here already you're in good shape.